storm-client/src/jvm/org/apache/storm/metrics2/TaskMetricRepo.java

```diff
@@ -34,23 +34,33 @@
     private SortedMap<String, Timer> timers = new TreeMap<>();
 
     public void addCounter(String name, Counter counter) {
-        counters.put(name, counter);
+        synchronized (this) {
+            counters.put(name, counter);
+        }
     }
 
     public void addGauge(String name, Gauge gauge) {
-        gauges.put(name, gauge);
+        synchronized (this) {
+            gauges.put(name, gauge);
+        }
     }
 
     public void addMeter(String name, Meter meter) {
-        meters.put(name, meter);
+        synchronized (this) {
```

Comment:
Maybe you explained this before. But why do they need to be `SortedMap`?

Reply:
I believe the reporters require a sorted map.  If we don't use one, we'd have to create one every report call.  We should add synchronization here too.

Reply:
In the reporter method, a list of new SortedMap is created if filter is not null.  Maybe we can create a dummy filter that returns true and use it when the filter parameter is null. In this way, `report` method will always report sortedMap.

And we use ConcurrentHashMap outside to reduce performance impact from synchronization.